EIPS/eip-7704.md

```diff
@@ -0,0 +1,137 @@
+
+The motivation behind the updated gas cost is to match the real cost of the operation, which is equivalent to an `SLOAD`.
+
+## Specification
+
+| Parameter                 | Value                                        |
+| ------------------------- | -------------------------------------------- |
+| `FORK_TIMESTAMP`          | TBD                                          |
+| `HISTORY_STORAGE_ADDRESS` | `0x25a219378dad9b3503c8268c9ca836a52427a4fb` |
+| `BLOCKHASH_SERVE_WINDOW`  | `256`                                        |
+
+The `BLOCKHASH` opcode semantics remains the same as before. From the `fork_block` (first block where `block.timestamp >= FORK_TIMESTAMP`), the `BLOCKHASH` instruction should be updated to retrieve the requested blockhash from the contract storage.
+
+### Contract Implementation
+
+Exact evm assembly that can be used for the contract to resolve `BLOCKHASH`
+
+```
+// check if input > 8 byte value and revert if this isn't the case
+// the check is performed by comparing the biggest 8 byte number with
+// the call data, which is a right-padded 32 byte number.
+push8 0xffffffffffffffff
+push0
+calldataload
+gt
+push1 0x39
+jumpi
+
+// check if input > blocknumber-1 then return 0
+push1 0x1
+number
+sub
+push0
+calldataload
+gt
+push1 0x31
+jumpi
+
+// check if blocknumber > input + 8192 then return 0, no overflow expected for input of 8 bytes
+push0
+calldataload
+push2 0x2000
+add
+number
+gt
+push1 0x31
+jumpi
+
+// mod 8192 and sload
+push2 0x1FFF
+push0
+calldataload
+and
+sload
+
+// load into mem and return 32 bytes
+push0
+mstore
+push1 0x20
+push0
+return
+
+// return 0
+jumpdest
+push0
+push0
+mstore
+push1 0x20
+push0
+return
+
+// revert
+jumpdest
+push0
+push0
+revert
+
+stop
+```
+
+Note that the input contract read `32` bytes input as `calldataload`. Users and clients doing EVM call to this contract should left pad the `arg` correctly.
+
+<!-- TODO: bytecode is based off on first version and will be updated once assembley is locked down as it changes contract sender and address -->
+
+Corresponding bytecode:
+`60203611603157600143035f35116029575f356120000143116029576120005f3506545f5260205ff35b5f5f5260205ff35b5f5ffd00`
+
+#### Deployment
+
+For the verkle_block (blocks where `block.timestamp >= FORK_TIMESTAMP`), we consider that the fork (`FORK_TIMESTAMP`) is already in effect for at least `HISTORY_SERVE_WINDOW` blocks (or from genesis if `fork_block < HISTORY_SERVE_WINDOW`). The following changes take effect:
+
+- `BLOCKHASH` MUST retrieve blockhashes from the system contract storage
```

Comment:
Is this really a MUST?  Where you get the data is irrelevant as long as the side effects are the same.  Clients may have a more efficient storage of recent block hashes that don't involve diving into storage, and then batch the trie paths into an accumulator to mark as read later.  

I think the spec would be improved by dropping this point, and moving an expanded point #3 to the fist point.

My main issue is with the word "retrieve" which I read as side-effect free where in this case the side effect of recording a read from storage is essential.  Perhaps making it a SHOULD and include language that if it is retrieved by other means that the side effects of a storage read should still be generated, but recording side effects seems to be the essence of the third point.

Reply:
Good Q. I agree that it does not make sense to require actual reads, we just need to data to be correct + the side effects (gas cost + witness) to be consistent. Will update

Reply:
adjusted

Reply:
have made things more clear now